### PR TITLE
Allow passing a list of Files to CustomTarget. Closes #1720

### LINKED ISSUE #1720: Allow passing a list of File objects to custom_target
Currently the following code fails with a python assertion about unknown type:

```meson
foo = custom_target(
  'foo',
  input : [files('foo'), files('bar')],
  output : 'foo.h',
  command : [my_prog, '@INPUT0@', '@INTPUT1@', '-o', '@OUTPUT@'],
)
```
However, replacing `[files('foo'), files('bar')]` with `files('foo') + files('bar')` does work. So this is probably pretty trivial to fix.

I'll have a look at it soon if on one else gets to it before then.

### PR DESCRIPTION
**Pull request by dcbaker**

### CODE REVIEW

File: test cases/common/149 list of file sources/meson.build [hunk 1]
```diff
@@ -0,0 +1,11 @@
+project('test', 'c')
+
+python = find_program('python3')
```
Comment (ignatenkobrain) [inline test cases/common/149 list of file sources/meson.build]: use python module instead

File: mesonbuild/build.py [hunk 2]
```diff
@@ -1417,6 +1417,8 @@ def process_kwargs(self, kwargs):
         self.sources = kwargs.get('input', [])
         if not isinstance(self.sources, list):
             self.sources = [self.sources]
+        else:
+            self.sources = flatten(self.sources)
```
Comment (nirbheek) [inline mesonbuild/build.py]: You can replace all this with:

```python
self.sources = flatten(kwargs.get('input, []))
```

**Comment by dcbaker** [comment]
That actually doesn't work input isn't a list. We could wrap `kwargs.get` in another list if you prefer the one liner, or we could go back to the if else. I'm doing the latter for the moment, since it seems like the less surprising of the two.

**Comment by dcbaker** [comment]
Or we could add a new function that make a one level deep list.

**Comment by dcbaker** [comment]
Or maybe flatten is wrong? given `a` is not a list, `flatten` will return `a`, but should it actually return `[a]`?

**Comment by nirbheek** [comment]
Hmm, we could make flatten always return a list, but I wonder if that will break something else. It shouldn't break anything though because in theory anything that would use flatten() should be prepared to have it return a list. Could you give that a shot?

**Comment by dcbaker** [comment]
Sure thing.

**Comment by dcbaker** [comment]
It doesn't seem to break anything in the test suite or in mesa to always return a list.

**Pull request merged**